docs/general-concepts/javascript/adding-javascript.md

```diff
@@ -0,0 +1,270 @@
+
+**type** - this is "script" for JavaScript
+
+**uri** - this tells Joomla where to find your code. It's the path under the media folder, **but with the js segment removed**. 
+Note that you can use minified JavaScript, as in the "myjs2" example.
+If you include both myjs2.js and myjs2.min.js in the js folder, 
+then if you switch on Joomla Debug in the Global Configuration the Web Asset Manager will include the non-minified version. 
+
+**attributes** - these are attributes which just get mapped to attributes of the HTML `<script>` tag within the page HTML output.
+
+**dependencies** - the name or names of other assets upon which your code depends, for example in the json above:
+- "com_example.myjs1" is dependent upon Joomla "core" (specified in media/system/joomla.asset.json as the Joomla core.js library)
+and upon "jquery" (specified in media/vendor/joomla.asset.json)
+- "com_example.myjs2" is dependent upon "form.validate" (specified in media/system/joomla.asset.json, for performing form field validation)
+- "com_example.myjs2" is also dependent upon "com_example.myjs1" (in the same joomla.asset.json file),
+so when myjs2.js is loaded Joomla will also load myjs1.js, and all of the dependencies of "com_example.myjs1" as well. 
+
+The Web Asset Manager orders the `<script>` tags in the HTML to ensure that the browser processes the dependencies first.
+
+How do you know the names of the Joomla libraries to include as dependencies? 
+Entries in the [Joomla JavaScript Library](./js-library/index.md) may help, but you may need to look through the code files
+- Joomla library assets are in media/system/joomla.asset.json, and refer to code in media/system/js
+- vendor library assets are in media/vendor/joomla.asset.json, and refer to code in media/vendor/js
+
+These two joomla.asset.json files are always processed by Joomla, so their assets are always available.
+
+**version** - the version you want to apply to your JavaScript code. 
+When you update your code then if you increase the version number then Joomla will force the browser to reload your code, instead of using a cached copy.
+(It does this by adding your version as a query parameter in the URL of your js file).
+
+You also need to get Joomla to copy your joomla.asset.json file into the media folder:
+
+```xml title="Extension manifest file"
+    <media destination="com_example" folder="media">
+        <filename>joomla.asset.json</filename>
+        <folder>js</folder>
+    </media>
+```
+
+### 3 Get Joomla to process your joomla.asset.json file
+
+If your extension is a component or a template then Joomla performs this step automatically for you, and you have nothing to do.
+
+If your extension is a plugin or a module then you need to register your file with the Web Asset Manager in your PHP code:
+
+```php
+$app = Factory::getApplication();
+$document = $app->getDocument();
```

Comment:
Since you refer to a plugin you don't need to use the super global here (we should try to avoid using Factory::).
In the Plugin you can use `$this->getApplication();` even better would be in the own plugin to set the document object it self but that's out of scope here.

Reply:
For general examples I think it is fine with `Factory::getApplication()`.
Only `$this->getApplication();` will be confusing for newbies, who still not fully know the context.

Maybe  just add extra comments, that:
- in module it available as `$app->getDocument()->getWebAssetManager()`;
- in plugin `$this->getApplication()->getDocument()->getWebAssetManager();` **but only when application is injected in service provider**;
- in component view `$this->document->getWebAssetManager()`

It may be obvious for us, but not for newbies.

Reply:
the reason why I mention it here, is because people do copy paste, which means reference to `Factory` which is wrong. The documentation have todo it right. If not here where else.

Reply:
I understood. I just wanted to say, when someone copy `$this->getApplication()` without context they will ended up with broken code. It is not guaranteed to be available in Plugin and it is different in Module and Views.

Reply:
I added a note about getting to the class instances in another way.